filters/ema_filter.cpp

```diff
@@ -0,0 +1,105 @@
+#include "ema_filter.hpp"
+
+#include <stout/foreach.hpp>
+
+#include <utility>
+
+namespace mesos {
+namespace serenity {
+
+
+Try<Nothing> EMAFilter::consume(const ResourceUsage& in) {
+  std::unique_ptr<ExecutorSet> newSamples(new ExecutorSet());
+  ResourceUsage product;
+
+  foreach(ResourceUsage_Executor inExec, in.executors()) {
+    if (!inExec.has_statistics() || !inExec.has_executor_info()) {
+      LOG(ERROR) << "Executor "
+                 << inExec.executor_info().executor_id().value()
+                 << " has not proper statistics or executor_info";
+    }
+    newSamples->insert(inExec);
+
+    // Check if EMA for given executor exists.
+    auto emaSample = this->emaSamples->find(inExec.executor_info());
+    if (emaSample == this->emaSamples->end()) {
+      // If not insert new one.
+      ExponentialMovingAverage ema(alpha);
+      emaSamples->insert(std::pair<ExecutorInfo, ExponentialMovingAverage>(
+          inExec.executor_info(), ema));
+
+    } else {
+      // Check if previousSample for given executor exists.
+      auto previousSample = this->previousSamples->find(inExec);
+      if (previousSample != this->previousSamples->end()) {
+        ResourceUsage_Executor* outExec = new ResourceUsage_Executor();
```

Comment:
Where is outExec  is cleared in case of Error?
Are you sure this code from here to end of "for" does not throw exceptions?